cpp/src/arrow/status.h

```diff
@@ -94,8 +94,14 @@ class ARROW_MUST_USE_RESULT ARROW_EXPORT Status;
 class ARROW_EXPORT Status {
  public:
   // Create a success status.
-  Status() : state_(NULL) {}
-  ~Status() { delete state_; }
+  Status() noexcept : state_(NULL) {}
+  ~Status() noexcept {
+    // ARROW-2400: On certain compilers, splitting off the slow path improves
+    // performance significantly.
+    if (ARROW_PREDICT_FALSE(state_ != NULL)) {
```

Comment:
Can we write this without the double negative? Does the branch prediction provide a material speedup? Or is it mostly from avoiding the destruction code + `noexcept`?

Reply:
Branch prediction is done by the CPU. Here we instruct the compiler that the `state_ != NULL` path is a cold path.

I'll remove the double negative.

Reply:
To answer your question, the speedup seems to stem from separating the fast path from the cold path.

Reply:
Hmm, removing the double negative makes the condition less readable IMHO (`!ARROW_PREDICT_TRUE(state_== NULL)`).

Reply:
Cool, leaving it as is fine by me.